src/consensus/tx_verify.cpp

```diff
@@ -212,26 +215,48 @@ bool Consensus::CheckTxInputs(const CTransaction &tx, CValidationState &state, c
     CAmount nValueIn = 0;
     CAmount nFees = 0;
     int nSpendHeight = -1;
-    for (unsigned int i = 0; i < tx.vin.size(); i++)
     {
-        const COutPoint &prevout = tx.vin[i].prevout;
-        const Coin &coin = inputs.AccessCoin(prevout);
-        assert(!coin.IsSpent());
-
-        // If prev is coinbase, check that it's matured
-        if (coin.IsCoinBase())
+        LOCK(inputs.cs_utxo);
+        for (unsigned int i = 0; i < tx.vin.size(); i++)
         {
-            if (nSpendHeight == -1)
-                nSpendHeight = GetSpendHeight(inputs);
-            if (nSpendHeight - coin.nHeight < COINBASE_MATURITY)
-                return state.Invalid(false, REJECT_INVALID, "bad-txns-premature-spend-of-coinbase",
-                    strprintf("tried to spend coinbase at depth %d", nSpendHeight - coin.nHeight));
+            const COutPoint &prevout = tx.vin[i].prevout;
+            const Coin &coin = inputs.AccessCoin(prevout);
+            assert(!coin.IsSpent());
+
+            // If prev is coinbase, check that it's matured
+            if (coin.IsCoinBase())
+            {
+                CAmount nCoinOutValue = coin.out.nValue;
+                int nCoinHeight = coin.nHeight;
+
+                // If there are multiple coinbase spends we still only need to get the spend height once.
+                if (nSpendHeight == -1)
+                {
+                    // GetSpendHeight requires cs_main but in order to maintain proper locking order cs_utxo
+                    // must be taken after cs_main.
+                    LEAVE_CRITICAL_SECTION(inputs.cs_utxo);
```

Comment:
Leave and re-enter is a bit of a lie because the locked variables cannot be used after inputs.cs_utxo is given.  So I would like to discourage its use.  In this case, the code correctly does not use coin.  However, this code would be a lot more clear if the lock just protected pulling the relevant data out of "coin" and into local variables.  This is already basically happening in lines 229 and 230, so a bit of reorg (pull these locals out of the if, assign them, and use in the else as well) could make this code both cleaner and lock cs_utxo for less time.

Reply:
We could do the copy for all checks and that would be cleaner, but here I was trying to avoid the copy on most iterations because coinbase inputs are relatively rare.